plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/procedure/RegisterTableProcedure.java

```diff
@@ -0,0 +1,234 @@
+        this.classLoader = getClass().getClassLoader();
+        this.registerTableProcedureEnabled = requireNonNull(icebergConfig, "icebergConfig is null").isRegisterTableProcedureEnabled();
+    }
+
+    @Override
+    public Procedure get()
+    {
+        return new Procedure(
+                SYSTEM_SCHEMA,
+                PROCEDURE_NAME,
+                ImmutableList.of(
+                        new Procedure.Argument(SCHEMA_NAME, VARCHAR),
+                        new Procedure.Argument(TABLE_NAME, VARCHAR),
+                        new Procedure.Argument(TABLE_LOCATION, VARCHAR),
+                        new Procedure.Argument(METADATA_FILE_NAME, VARCHAR, false, null)),
+                REGISTER_TABLE.bindTo(this));
+    }
+
+    public void registerTable(
+            ConnectorSession clientSession,
+            String schemaName,
+            String tableName,
+            String tableLocation,
+            String metadataFileName)
+    {
+        try (ThreadContextClassLoader ignored = new ThreadContextClassLoader(classLoader)) {
+            doRegisterTable(
+                    clientSession,
+                    schemaName,
+                    tableName,
+                    tableLocation,
+                    Optional.ofNullable(metadataFileName));
+        }
+    }
+
+    private void doRegisterTable(
+            ConnectorSession clientSession,
+            String schemaName,
+            String tableName,
+            String tableLocation,
+            Optional<String> metadataFileName)
+    {
+        if (!registerTableProcedureEnabled) {
+            throw new TrinoException(PERMISSION_DENIED, "register_table procedure is disabled");
+        }
+        checkProcedureArgument(schemaName != null && !schemaName.isEmpty(), "schema_name cannot be null or empty");
+        checkProcedureArgument(tableName != null && !tableName.isEmpty(), "table_name cannot be null or empty");
+        checkProcedureArgument(tableLocation != null && !tableLocation.isEmpty(), "table_location cannot be null or empty");
+        if (metadataFileName.isPresent()) {
+            validateMetadataFileName(metadataFileName.get());
+        }
+        SchemaTableName schemaTableName = new SchemaTableName(schemaName, tableName);
+        TrinoCatalog catalog = catalogFactory.create(clientSession.getIdentity());
+        if (!catalog.namespaceExists(clientSession, schemaTableName.getSchemaName())) {
+            throw new TrinoException(SCHEMA_NOT_FOUND, format("Schema '%s' does not exist", schemaTableName.getSchemaName()));
+        }
+
+        TrinoFileSystem fileSystem = fileSystemFactory.create(clientSession);
+        String metadataLocation = getMetadataLocation(fileSystem, tableLocation, metadataFileName);
+        validateLocation(fileSystem, metadataLocation);
+        try {
+            // Try to read the metadata file. Invalid metadata file will throw the exception.
+            TableMetadataParser.read(fileSystem.toFileIo(), metadataLocation);
```

Comment:
@findepi | @alexjo2144 | @findinpath, Need your view over here about the invalid metadata file.

Reply:
Reference to https://github.com/trinodb/trino/pull/14375#discussion_r1014628366

Reply:
IMO it would be better to fail fast when registering the table and not wait until the table is first queried.
Note that the two operations -  registering the table and  querying the table - can very well be a few days apart from each other and by the time the query on the table fails, the person who has done the registering may have lost context related to the `register_table` operation.